File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        long prev=1e11-1;
        return help(root,prev);

    }
    bool help(TreeNode* root,long& prev){
        if(root==NULL)return true;
        bool l=help(root->left,prev);
        if(prev!=(1e11-1)&&prev>=root->val)return false;
        prev=root->val;
        bool r=help(root->right,prev);
        return l&&r;
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        prev=1e11-1
        def help(root) -> bool:
            nonlocal prev
            if not root: return True
            l=help(root.left)
            if prev!=1e11-1 and prev >= root.val: return False
            prev=root.val
            r=help(root.right)
            return l and r
        return help(root)